msgpack-core/src/main/java/org/msgpack/core/MessageUnpacker.java

```diff
@@ -531,57 +525,143 @@ public void skipValue() throws IOException {
      */
     private static MessageTypeException unexpected(String expected, byte b)
             throws MessageTypeException {
-        ValueType type = ValueType.valueOf(b);
-        return new MessageTypeException(String.format("Expected %s, but got %s (%02x)", expected, type.toTypeName(), b));
+        MessageFormat format = MessageFormat.valueOf(b);
+        String typeName;
+        if (format == MessageFormat.NEVER_USED) {
+            typeName = "NeverUsed";
+        } else {
+            String name = format.getValueType().name();
+            typeName = name.substring(0, 1) + name.substring(1).toLowerCase();
+        }
+        return new MessageTypeException(String.format("Expected %s, but got %s (%02x)", expected, typeName, b));
     }
 
-    public MessageFormat unpackValue(ValueHolder holder) throws IOException {
+    public ImmutableValue unpackValue() throws IOException {
         MessageFormat mf = getNextFormat();
         switch(mf.getValueType()) {
             case NIL:
                 unpackNil();
-                holder.setNil();
-                break;
+                return ValueFactory.newNilValue();
             case BOOLEAN:
-                holder.setBoolean(unpackBoolean());
-                break;
-            case INTEGER: {
-                unpackInteger(holder.getIntegerHolder());
-                holder.setToInteger();
-                break;
+                return ValueFactory.newBooleanValue(unpackBoolean());
+            case INTEGER:
+                switch (mf) {
+                    case UINT64:
+                        return ValueFactory.newIntegerValue(unpackBigInteger());
+                    default:
+                        return ValueFactory.newIntegerValue(unpackLong());
+                }
+            case FLOAT:
+                return ValueFactory.newFloatValue(unpackDouble());
```

Comment:
Discussed this topic with @frsyuki. MessagePack type system will not preserve the error range of float values since some msgpack implementation cannot have float values. So application developers needs to be careful so as not to write a code that depends on float precision. Even if the data is written in FLOAT32, Value interface will represent it in double precision.